completion_aggregator/core.py

```diff
@@ -0,0 +1,322 @@
+"""
+Core aggregator functionality.
+
+This is the engine that takes BlockCompletion objects and
+converts them to Aggregators.
+"""
+
+import logging
+from collections import namedtuple
+from datetime import datetime
+
+import pytz
+import six
+from xblock.completable import XBlockCompletionMode
+from xblock.core import XBlock
+
+from django.utils import timezone
+
+from . import compat
+from .cachegroup import CacheGroup
+from .models import Aggregator, StaleCompletion
+from .utils import BagOfHolding
+
+OLD_DATETIME = pytz.utc.localize(datetime(1900, 1, 1, 0, 0, 0))
+UPDATER_CACHE_TIMEOUT = 600  # 10 minutes
+
+CompletionStats = namedtuple('CompletionStats', ['earned', 'possible', 'last_modified'])
+
+log = logging.getLogger(__name__)
+
+
+class UpdaterCache(object):
+    """
+    Cache course blocks for Aggregation Updater.
+
+    Uses `completion_aggregator.cachegroup.CacheGroup` to allow bulk
+    invalidation by course_key.
+    """
+
+    cache_key_template = "completion_aggregator.updater.{user_id}-{course_key}-{root_block}"
+
+    def __init__(self, user_id, course_key, root_block):
+        """
+        Create a new Updater Cache for the provided user_id, course_key, and root_block.
+        """
+        self.user_id = user_id
+        self.course_key = course_key
+        self.root_block = root_block
+
+    def get(self):
+        """
+        Return the cached BlockStructure for the current cache entry.
+        """
+        return CacheGroup().get(self.cache_key)
+
+    def set(self, value):
+        """
+        Cache a BlockStructure for the current cache entry.
+
+        Sets the group to `str(self.course_key)`.
+        """
+        group = six.text_type(self.course_key)
+        CacheGroup().set(group, self.cache_key, value, timeout=UPDATER_CACHE_TIMEOUT)
+
+    @property
+    def cache_key(self):
+        """
+        Create a key to identify the current cache entry.
+        """
+        return self.cache_key_template.format(
+            user_id=self.user_id,
+            course_key=self.course_key,
+            root_block=self.root_block or 'COURSE',
+        )
+
+
+class AggregationUpdater(object):
+    """
+    Class to update aggregators for a given course and user.
+    """
+
+    def __init__(self, user, course_key, modulestore, root_block=None):
+        """
+        Create an aggregation updater for the given user and course.
+
+        Also takes a modulestore instance.
+        """
+        self.user = user
+        self.course_key = course_key
+        self.raw_root_block = root_block
+        self.cache = UpdaterCache(self.user.id, self.course_key, self.raw_root_block)
+
+        cache_entry = self.cache.get()
+        if cache_entry:
+            self.course_blocks = cache_entry
```

Comment:
``self.root_block`` is no longer set if there is a cache entry, which causes an error to be thrown in ``calculate_updated_aggregators``:

```python
Traceback (most recent call last):
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/core/handlers/base.py", line 132, in get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/utils/decorators.py", line 145, in inner
    return func(*args, **kwargs)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/views/decorators/csrf.py", line 58, in wrapped_view
    return view_func(*args, **kwargs)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/views/generic/base.py", line 71, in view
    return self.dispatch(request, *args, **kwargs)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/rest_framework/views.py", line 466, in dispatch
    response = self.handle_exception(exc)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/rest_framework/views.py", line 463, in dispatch
    response = handler(request, *args, **kwargs)
  File "/edx/src/openedx-completion-aggregator/completion_aggregator/api/v1/views.py", line 408, in get
    ) for enrollment in paginated
  File "/edx/src/openedx-completion-aggregator/completion_aggregator/serializers.py", line 120, in __init__
    self.update_aggregators(aggregators or [], root_block=root_block, is_stale=is_stale)
  File "/edx/src/openedx-completion-aggregator/completion_aggregator/serializers.py", line 155, in update_aggregators
    force=True,
  File "/edx/src/openedx-completion-aggregator/completion_aggregator/core.py", line 293, in calculate_updated_aggregators
    return updater.calculate_updated_aggregators(changed_blocks, force)
  File "/edx/src/openedx-completion-aggregator/completion_aggregator/core.py", line 148, in calculate_updated_aggregators
    self.update_for_block(self.root_block, affected_aggregators, force)
AttributeError: 'AggregationUpdater' object has no attribute 'root_block'
```

Reply:
Oh yep.  I had some other information in the cache, and then I pulled it out.  Looks like I overdid it.  It's fixed now.  Please give it another look.